toffy/streak_detection.py

```diff
@@ -0,0 +1,409 @@
+        data for streak correction.
+        input_image (np.ndarray): The input image (Numpy array representation of the tiff file)
+        to be used for streak detection.
+
+    Returns:
+        np.ndarray: The corrected image.
+    """
+    # Pad the image for edge cases.
+    padded_image = np.pad(input_image.copy(), pad_width=(1, 1), mode="edge")
+    corrected_image = padded_image.copy()
+    # Correct each streak
+    for region in streak_data.filtered_streak_df.itertuples():
+        corrected_image[region.max_row, region.min_col : region.max_col] = _mean_correction(
+            padded_image,
+            region.min_row,
+            region.max_row,
+            region.min_col,
+            region.max_col,
+        )
+    # Crop and return the 'unpadded' image.
+    return util.crop(corrected_image, crop_width=(1, 1), copy=True)
+
+
+def _mean_correction(
+    input_image: np.ndarray, min_row: int, max_row: int, min_col: int, max_col: int
+) -> np.ndarray:
+    """Performs streak-wise correction by: setting the value of each pixel in the streak to the
+    mean of pixel above and below it.
+
+    Args:
+        input_image (np.ndarray): The input image (Numpy array representation of the tiff file) to
+        be used for streak correction.
+        min_row (int): The minimum row index of the streak. The y location where the streak
+        starts.
+        max_row (int): The maximum row index of the streak. The y location where the streak ends.
+        min_col (int): The minimum column index of the streak. The x location where the streak
+        starts.
+        max_col (int): The maximum column index of the streak. The x location where the streak
+        ends.
+
+    Returns:
+        np.ndarray: Returns the corrected streak.
+    """
+    streak_corrected: np.ndarray = np.mean(
+        [
+            # Row above
+            input_image[min_row, min_col + 1 : max_col + 1],
+            # Row below
+            input_image[max_row + 1, min_col + 1 : max_col + 1],
+        ],
+        axis=0,
+        dtype=np.int8,
+    )
+
+    return streak_corrected
+
+
+def streak_correction(
+    fov_dir: Union[str, Path],
+    image_name: str = "Noodle",
+    mask_data: bool = False,
+) -> Optional[StreakData]:
+    """Takes a `fov` directory and a user specified image for streak detection. Once all the
+    streaks have been detected on that image, they are corrected via an averaging method. The
+    function can also return a DataClass containing various binary masks and dataframes which were
+    used for filtering and correction when prompted to.
+
+    Args:
+        fov_dir (Union[str, Path]): A directory containing the fov and all it's channels for
+        correction.
+        image_name (str): The name of the image used for identifying and removing the streaks.
+        Defaults to "Noodle".
+        mask_data (bool): If `True`, returns a DataClass consisting of binary masks and dataframes
```

Comment:
Right now this function is self-contained and can load the image data, process it, and then save it. However, in the future we may have this streak detection be an automatic part of the pipeline. In that case, we may already have the data loaded. 

It would be more modular if you broke this functionality up. Lets have streak_correction take in an xarray of images, and return an xarray of images. The loading/saving can be pulled out into separate functions